interfaces/builtin/mount_control.go

```diff
@@ -368,21 +417,38 @@ func validateMountOptions(mountInfo *MountInfo) error {
 	if len(mountInfo.options) == 0 {
 		return errors.New(`mount-control "options" cannot be empty`)
 	}
-	// With "functionfs" none of the valid "options" can be harmful so no need to check against allowedMountOptions
-	if mountInfo.isType("functionfs") {
-		if err := validateNoAppArmorRegexpWithError(`cannot use mount-control "option" attribute`, mountInfo.options...); err != nil {
-			return err
+	if err := validateNoAppArmorRegexpWithError(`cannot use mount-control "option" attribute`, mountInfo.options...); err != nil {
+		return err
+	}
+	var types []string
+	if mountInfo.hasType() {
+		fsExclusiveOption := optionIncompatibleWithFsType(mountInfo.options)
+		if fsExclusiveOption != "" {
+			return fmt.Errorf(`mount-control option %q is incompatible with specifying filesystem type`, fsExclusiveOption)
 		}
+		types = mountInfo.types
 	} else {
-		for _, o := range mountInfo.options {
-			if !strutil.ListContains(allowedMountOptions, o) {
-				return fmt.Errorf(`mount-control option unrecognized or forbidden: %q`, o)
+		types = defaultFSTypes
+	}
+	for _, o := range mountInfo.options {
+		if strutil.ListContains(allowedKernelMountOptions, o) {
+			continue
+		}
+		if strutil.ListContains(allowedUserspaceMountOptions, o) {
+			continue
```

Comment:
no test seems to reach here

Reply:
I've updated tests and I believe this should be covered now?